add C-lang/llvm based metaware toolchain support in Zephyr

vonhust opened the pull request:
This PR contains Synopsys' work on supporting C-lang/llvm based metaware toolchain support in Zephyr.

As Zephyr toolchain work group is kicked off, we share our work here and welcome review.

Now can we compile lots of zephyr examples and tests with metarware toolchain, but still a lot work of to do:
    * C++ support
    * compile options mapping from GCC to mwdt
    * binutils transfer.
    * corner cases

# update of 2020-03-05
Here is a summary of mwdt support in Zephyr..
## Current status

Now, most of non-USERSPACE samples can be built with MWDT and run in nsim. For USERSPACE samples, because of address alignments requirement, there are a lot of issues for MPUv2 configurations. The result of MPUv3 (nsim_sem) is better than MPUv2, as v3’s requirement of address alignment is weaker.

## Background

MWDT is SNPS’s new generation LLVM/C-lang based toolchain. Although Zephyr already has some support of llvm/c-lang toolchain, but MWDT has different interface (compile options), linker, assembler. Therefore, the exist llvm/c-lang support in zephyr cannot be reused for MWDT, everything must be done from the scratch.

The work can be divided into 4 parts:

## PART I build system

   Zephyr build system is cmake based with lots of python tools. The general flow is in https://docs.zephyrproject.org/latest/guides/build/index.html. The goal of this part is to pass the configuration of cmake. In this part, we need to set up the toolchain related cmake files for mwdt. GNU toolchain’s cmake files are good reference. The challenges of this part are

 * Basic understanding of zephyr’s build flow and knowledge of cmake
 * Compiler options check where “zephyr_check_compiler_flag(C "" toolchain_is_ok)” is very important. It requires the correct setting of CMAKE_REQUIRED_FLAGS
 * Generation of device tree related files. Execute_process @line 122 of dts.cmake will call c compiler to preprocess the dts files. It works for gcc, but failed for mwdt. As a workaround, host gcc can be used. The root cause of this issue may be in the assembler or preprocessor of mwdt.

## PART II: compilation
The goal of this part is to compile the source code successfully

* Compile options. In Zephyr, lots of compile options are used, covering generic options, optimization, path, processor specific, etc. This compile options are GNU centric. It requires mapping from gnu compile options to mwdt compile options. But for some options, there are no corresponding mwdt options. I suggest Zephyr needs to list what options are critical, what are optional. Here are a list of options not supportted in MWDT
  * gcc specific coverage related options, this affect the coverage test of zephyr
  * no -std=c++14, -std=c++17 -std=c++2a
  * no -fsanitize=undefined
  * no mstack-protector-guard=global
  * no security_fortify related options

* Language feature, extension and builtins. The language (C/C++/Assembly) extensions used in Zephyr are GCC-centric. It requires a mapping between gcc and mwdt. For C/C++, luckily, as MWDT is llvm/c-lang based, so most of language feature, extension and builtin can be reused (there are still some corner cases); For assembly, more work is required. Because assembly is more low-level specific and architecture specific, it has few impact on application, API but may require lots of changes in exist assembly codes.

## PART III: Linkage

The goal of this part to link the compiled object files, archive files into the final elf file.

MWDT’s linker and link script is fully different with gnu’s. Compared with gnu’s linker, the features of MWDT’s linker and ld script are limited. Meanwhile, the linker script used by Zephyr is very complex, has lots of symbols and sections, especially when USER SPACE is configured. The linkage support is the most challenging work in MWDT support. Here are the challenges:

* Map Zephyr’s gnu linker script into MWDT’s linker script
* Lots of GNU linker primitives are not supported by MWDT, including, SORT, SORT_BY_NAME, operation on PC counter (.), LOG2CEIL (which is required by the power of 2 alignment of ARC MPU v2), ASSERT
 * Lots of symbol are gnu specific, e.g. C++

What’s more, even the final elf file is linked. It may be a wrong result because of address alignment, gapping etc.

## PART IV: binutils

The goal of this part is to generate different kind of output files(.bin, .hex, .S record), dump symbols or sections from the elf which may be used by 2nd linkage, mem footprint, etc.

Compared with MWDT, the features of gnu binutils, are rich and strong, especially the objcopy cmd. Required by USERSPACE feature, section rename of objcopy is used. But we can’t do section rename with MWDT, so still rely on using arc gnu’s objcopy to process the files generated by MWDT.

Lots of python tools in Zephyr are designed based on gnu binutils. This means these tools cannot be directly used by MWDT.

# update of 2020-03-19
  * can run sanitycheck with metaware toolchain
  * testes can be built with metaware toolchain (build only)
      * tests/kernel, covering USERSPACE
  * testes can pass with metaware toolchain
      * tests/kernel, without USERSPACE.
      * because ARC MPUv3 only requires 32 bytes alignment, so USERSPACE with ARC MPUv3 also can pass
      * suggestions for toolchain tests
         * STEP1 : hello world
         * STEP2 : philosopher for context switch and systick
         * STEP3 : tests/kernel  without USERSPACE for core kernel features
         * STEP4 : tests/kernel with USERSPACE  for complete kernel features

 # update of 2020-09-04
  * significant code cleanup
  * drop the parts which have been already merged (like replacing ASM code with C, changes in generic python scripts, etc...)
  * use new metaware linker (LLDAC instead of LDAC) which is more feature-full. That allowed us do drop a lot of hacks done for LDAC due to some missing features.
  * try to avoid changes in generic code/scripts as much as possible
  * up to 95% of tests passing with metaware toolchain when testing on simulation platforms (nsim_em, nsim_hs, hsim_sem)

daor-oti commented on CMakeLists.txt:
```diff
@@ -291,10 +291,14 @@ toolchain_cc_warning_error_implicit_int()
 
+#
+# mwdt doesn't have the following options
+if(COMPILER STREQUAL gcc)
```
This check is not needed, as the zephyr_cc_option checks whether the given flag is supported by the compiler and in case it is not, it will not be added. Also, it may not only be gcc which needs these flags.

vonhust replied:
Thanks for explanation, i will update this part

daor-oti commented on CMakeLists.txt:
```diff
@@ -318,7 +323,10 @@ if(NOT CMAKE_C_COMPILER_ID STREQUAL "Clang")
 endif()
 
+# no this option for mwdt
+if(COMPILER STREQUAL gcc)
```
Same comment as above, check not needed as the zephyr_cc_option_ifdef already verifies compiler flag support.

mped-oticon replied:
Indeed, many compilers (not only gcc) may support -fstack-usage.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -353,6 +364,7 @@ get_property(TOPT GLOBAL PROPERTY TOPT)
 set_ifndef(  TOPT -Wl,-T) # clang doesn't pick -T for some reason and complains,
                           # while -Wl,-T works for both, gcc and clang
 
+# mwdt should have custom linker script, or do some speical handling in linker.ld
```
mwdt? Is this a TODO?

vonhust replied:
it's a hint for me, alike TODO,  i will clean them,

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -753,7 +765,7 @@ if(CONFIG_CODE_DATA_RELOCATION)
   toolchain_ld_relocation()
 endif()
-
+# mwdt has no -ftest-coverage -fprofile-arcs
```
Is this a TODO?

SebastianBoe commented on cmake/compiler/mwdt/target.cmake:
```diff
@@ -0,0 +1,165 @@
+  -c
+  -Hnosdata
+  -Hnolib
+  -Hnocrt
```
-Hnocrt

isn't this what

-fno-freestanding

is?

vonhust replied:
In Metaware, -Hnocrt is an option for linker, means donot link toolchain provided startup files.
It's similar to -nostartup

SebastianBoe replied:
This is OK then

SebastianBoe commented on cmake/compiler/mwdt/target.cmake:
```diff
@@ -0,0 +1,165 @@
+
+macro(toolchain_cc_security_canaries)
+  zephyr_compile_options(-fstack-protector-all)
+  message(WARNING "no support of -mstack-protector-guard=global")
```
Drop this warning.

SebastianBoe commented on cmake/compiler/mwdt/target.cmake:
```diff
@@ -0,0 +1,165 @@
+
+
+macro(toolchain_cc_security_fortify)
+  message(WARNING "no support of _FORTIFY_SOURCE")
```
Drop this warning.

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -120,7 +120,7 @@ if(SUPPORTS_DTS)
   # so that changes to DT sources are detected.
   execute_process(
-    COMMAND ${CMAKE_C_COMPILER}
+    COMMAND gcc
```
This is not OK.

If this compiler can't preprocess DT sources then you need to introduce an abstraction like

"CMAKE_C_COMPILER_fully_featured"

or similar that can preprocess DT sources.

Then normally we can set this variable to CMAKE_C_COMPILER, and as an exception do something else for
this compiler. But the best thing would be to have this compiler be able to preprocess DTS.

vonhust replied:
It may be a bug of metaware compiler,  the options are ok, but emitt error msg, i am checking with snps compiler team.

SebastianBoe commented on cmake/linker/mwdt/target.cmake:
```diff
@@ -0,0 +1,203 @@
+# SPDX-License-Identifier: Apache-2.0
+
+find_program(CMAKE_LINKER ${CROSS_COMPILE}ldac PATH ${TOOLCHAIN_HOME} NO_DEFAULT_PATH)
```
It's non-standard to detect the linker this late isn't it?

Let's align with the other toolchains.

SebastianBoe commented on cmake/linker/mwdt/target.cmake:
```diff
@@ -0,0 +1,203 @@
+    ${extra_dependencies}
+    # NB: 'linker_script_dep' will use a keyword that ends 'DEPENDS'
+    ${linker_script_dep}
+    COMMAND gcc
```
This is not OK either, see comment about preprocessing DT.

SebastianBoe commented on cmake/linker/mwdt/target.cmake:
```diff
@@ -0,0 +1,203 @@
+endfunction(toolchain_ld_link_elf)
+
+macro(toolchain_ld_baremetal)
+# LINKERFLAGPREFIX comes from linker/ld/target.cmake
```
Does it? But don't we need it to be

`-Hldopt=`

SebastianBoe commented on cmake/linker/mwdt/target.cmake:
```diff
@@ -0,0 +1,203 @@
+    -Hnocopyr
+    -Hnosdata
+    -Hnocrt
+    -Xtimer0 # to depress the warning message
```
to suppress the warning message, we wouldn't want to depress it.

vonhust replied:
-Xtimer0 in fact is a cpu related option,  mwdt requires at least one -Xxxx option, or warn msg will come out , so i add -Xtimer 0 here which is a common option for alll arc processors.

SebastianBoe replied:
I'm sure it's correct to add -Xtimer 0 here, but I'm saying that the comment should be saying "suppress", not "depress".

SebastianBoe commented on include/linker/intlist.ld:
```diff
@@ -44,3 +45,18 @@ SECTION_PROLOGUE(.intList,,)
+#ifndef LINKER_PASS2
+SECTION_PROLOGUE(.intList,,)
+{
+	KEEP(*(.irq_info.*))
```
It's cleaner to just add .irq_info.* to the pre-existing list of input sections than to ifdef the whole thing.

It will be harmless to gcc.

But a comment explaining why it's .irq_info. on MWDT in the commit message is in order.

andyross replied:
Agreed.  Just in general this patch has a TON of cut/paste linker code intended to work around what look to me like fairly trivial syntax changes between GNU binutils and MWDT.  Surely we can find one syntax that works for both, and where we can't we should probably be investigating our own tooling (e.g. python scripts run between linker passes, etc...) that can be made portable.

What we absolutely don't want is a bunch of custom code in our core infrastructure that can only be exercised by people using a proprietary toolchain for a single platform.  This code needs to run unmodified everywhere so that everyone has the ability to modify the linker framework.

SebastianBoe commented on include/linker/kobject-text.ld:
```diff
@@ -22,6 +24,7 @@
 #endif
 #endif
 
+#if defined(__GCC_LINKER_CMD__)
```
It's better to use feature-flags.

e.g.

ifdef SUPPORTS_ASSERT

or ifdef SUPPORTS_MAX,

depending on what mwdt was struggling with.

SebastianBoe commented on include/linker/linker-defs.h:
```diff
@@ -51,11 +51,22 @@
+		__device_##level##_start = .;			\
+		KEEP(*(MWDT_STRINGIFY(.init_##level[0-9].*)));		\
+		KEEP(*(MWDT_STRINGIFY(.init_##level[1-9][0-9].*)));
+#else
```
Does MWDT not support SORT?

What is the problem here?

vonhust replied:
MWDT does not support SORT like ld.

SebastianBoe replied:
Does it happen to SORT by default? Because if not, then surely the devices will
be initialized in the wrong order and break?

SebastianBoe commented on include/linker/linker-defs.h:
```diff
@@ -73,9 +84,15 @@
 /* define a section for undefined device initialization levels */
+#if defined(__MWDT_LINKER_CMD__)
+#define DEVICE_INIT_UNDEFINED_SECTION()		\
+		KEEP(*(MWDT_STRINGIFY(.init_[_A-Z0-9]*)))
```
Doesn't the correctness depend on SORT? Does this work?

SebastianBoe commented on include/linker/linker-tool-mwdt.h:
```diff
@@ -0,0 +1,104 @@
+/*
+ * mwdt linker doesn't have the following directives
+ */
+#define SORT(x) x
```
Are things sorted by default?

How does this not break badly?

SebastianBoe commented on include/linker/sections.h:
```diff
@@ -45,8 +45,13 @@
 /* Interrupts */
+#if defined(__MWDT_LINKER_CMD__)
+#define IRQ_VECTOR_TABLE	.gnu.linkonce.irq_vector_table.*
+#define SW_ISR_TABLE		.gnu.linkonce.sw_isr_table.*
```
Do we know why mwdt is generating the . suffix and gcc is not?

SebastianBoe commented on kernel/init.c:
```diff
@@ -465,6 +465,7 @@ void z_early_boot_rand_get(u8_t *buf, size_t length)
 		u8_t *p_rndbits = (u8_t *)&rndbits;
 
 		rndbits = sys_rand32_get();
+		rndbits = 0;
```
?

SebastianBoe commented on include/arch/arc/v2/linker_mwdt.ld:
```diff
@@ -83,6 +83,7 @@ SECTIONS {
  */
 #include <snippets-text-start.ld>
 		*(".text")
+		*(".text*")
```
It's cleaner to do

*(.text .text* .text.*)

Then the linker is not forced to place these input sections after each other and has more freedom to optimize.

SebastianBoe commented on include/linker/linker-tool-mwdt.h:
```diff
@@ -21,6 +21,18 @@
 #define SORT(x) x
 #define ASSERT(x, y)
 #define SUBALIGN(x) ALIGN(x)
+#define MAX(x, y)
```
It's better to get an error than to silently remove code as we do here.

mped-oticon commented on CMakeLists.txt:
```diff
@@ -309,6 +313,7 @@ zephyr_compile_options(${COMPILER_OPT_AS_LIST})
 
 # TODO: Include arch compiler options at this point.
 
+# TODO: special handling for mwdt here, mwdt is Clang based
```
So mwdt is clang-based, meaning the parser.
But does it accept {all the same, a superset, or different} options as clang?
I'd suggest mwdt gets its own CMAKE_C_COMPILER_ID, unless it is completely commandline compatible with clang.

FYI, verbatim compiler options in the root CMakeLists is considered cheating :) Please consider using the abstractions already inplace.

vonhust replied:
it's clang based, but lots of different options to keep compatible with old metware compiler.

SebastianBoe commented:
> I suggest Zephyr needs to list what options are critical, what are optional. 

zephyr_compile_options() are critical,
zephyr_cc_options() are optional.

But there are definitely some misclassified flags.

SebastianBoe commented on include/arch/arc/arch.h:
```diff
@@ -206,6 +206,10 @@ extern "C" {
+#if defined(__CCAC__)
+/* fixme: no _Static_assert in mwdt, make it a empty macro */
+#define _ARCH_MEM_PARTITION_ALIGN_CHECK(start, size)
+#else
```
Could we redefine BUILD_ASSERT_MSG to be NOP instead?

Also, can we use feature-flags instead of vendor-flags. E.g. if defined(NO_STATIC_ASSERT) instead of if defined(__CCAC__).

tejlmand commented:
@vonhust Thanks for this.
I'm currently working on a draft for this: #16031 that I hope to post soon.
So let's discuss way forward, based on this PR, current design, and improved design.

nashif commented on include/toolchain/gcc.h:
```diff
@@ -274,7 +274,22 @@ do {                                                                    \
  * Need to use assembly macros because ';' is interpreted as the start of
  * a single line comment in the ARC assembler.
  */
+#if defined(__CCAC__)
```
you probably want to create your own metaware.gcc and add special handling in there.

andyross replied:
Or just drop this entirely.   This whole mess is actually pre-existing, and is an attempt to fit the pre-existing GTEXT()/GDATA() assembly API for ARM and x86 into the ARC assembly syntax.

But... who cares?  It's an assembly macro.  You only every use it from ARC assembly.  There's zero value to making this portable at all, especially if it incurs this kind of messy syntax cost.

And it's a silly, trivial macro anyway.  All it does is declare an externally linked symbol.  Most of our more recent platforms (c.f. x86_64, xtensa don't use it at all, aarch64 does but only sparingly).  Really we've been moving away from most of our early assembly frameworks in a lot of our kernel APIs.  Maybe we should just junk this.  Now, for arc, would be  a good time.

tejlmand commented:
@vonhust you can comment here https://github.com/zephyrproject-rtos/zephyr/pull/24851, and we might be able to work in a direction that makes it easier to add additional toolchain support.

abrodkin reviewed (changes requested):
First of all could you please rebase that work on top of the current master so that it's easier to review and some changes like MDB runner will be used from the existing upstream tree.

Then `%U1` is a valid modifier at least for ARC & GCC toolchain, see https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/issues/285. So blindly removal might not be the best idea. I'd rather propose to switch to a generic accessors written in pure C. Something along the lines to what we now have in U-Boot, see https://gitlab.denx.de/u-boot/u-boot/-/blob/master/arch/arc/include/asm/io.h. That is way more flexible, much easier to maintain and really portable.

IRISZZW commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -1739,8 +1739,8 @@ def run_cmake(self, args=[]):
             '-DEXTRA_CFLAGS="-Werror ',
-            '-DEXTRA_AFLAGS=-Wa,--fatal-warnings',
-            '-DEXTRA_LDFLAGS="{}'.format(ldflags),
+            # '-DEXTRA_AFLAGS=-Wa,--fatal-warnings',
```
I will undo this commit, and push it in our own repo, and find a better way to fix it.

andyross commented on tests/kernel/gen_isr_table/src/main.c:
```diff
@@ -136,8 +136,15 @@ void isr6(void *param)
  */
+#if defined(__CCAC__)
+/* MWDT does not support _attribute__((optimize("-O0"))), buf the following
+ * it equals -Os
```
This function is a confusing mess, and IMHO this makes it worse.  Really someone should come in there and implement the ordering constraints described in a portable way (e.g. call stuff through a C function pointer and not compiler voodoo, whatever).

But barring that, please make this a NO_OPTIMIZE macro or something and declare it at the toolchain level.

andyross reviewed (dismissed):
Saw the traffic on this and came by with some drive by comments about the changes to the general code.

The big complaint is all the linker mess.  It seems like metaware presents a linkage environment that is **almost** the same as the default GNU one, but instead of making the two work compatibly this has duplicated code all over the place and made tiny changes.  Let's pick a single set of rules and syntax that work.

Commit:
ARC: handle the difference of assembly macro definition

GNU toolchain and MWDT (Metware) toolchain have different style
for accessing arguments in assembly macro. Implement the
preprocessor macro to handle the difference.

Make all ASM macros in swap_macros.h compatible for both ARC
toolchains.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: make the assembly codes compatible

Make the assembly codes compatible with both GNU
and Metaware toolchain.

* replace ".balign" with ".align"
  ".align" assembler directive is supposed by all
  ARC toolchains and it is implemented in a same
  way across ARC toolchains.
* replace "mov_s __certain_reg" with "mov __certain_reg"
  Even though GCC encodes those mnemonics and even real
  HW executes them according to PRM these are restricted
  ones for mov_s and CCAC rightfully refuses to accept
  such mnemonics. So for compatibility and clarity sake
  we switch to 32-bit mov instruction which allows use
  of all those instructions.
* Add "%%" prefix while accessing registers from inline
  ASM as it is required by MWDT.
* Drop "@" prefix while accessing symbols (defined in C
  code) from ASM code as it is required by MWDT.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

/#

Commit:
ARC: rename arch_switch implementation to z_arc_switch

"arch_switch" is declared as an inline function in kswap.h,
it should be a wrapper of arch level switch. The difference
of declaration and implementation of "arch_swich" causes
warning from MWDT compiler.

Use "arch_switch" with proper declararion (which is just
wraper for "z_arc_switch") to do conext switch for ARC.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: change direct IRQ declaration for metaware toolchain

* change direct IRQ declaration for metaware toolchain
* drop unused irq-related definitions

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: add linker script template for metware toolchain

Add linker script template for MWDT toolchain (linker-mwdt.ld)
Move linker.ld to linker-gnu.ld (without changes)

The "linker.ld" is wraper now.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
cmake: add metaware toolchain support in build system

* add metaware toolchain support in build system:
   * compiler
   * linker
   * binutils
     * gcc objcopy is used because it can't be replaced
       with mwdt's binutils currently
* To use ARC metaware toolchain, you'd better:
   * in Linux/Unix environment
   * install arc gcc/zephyr toolchain to use gnu's objcopy
     tool
   * set ZEPHYR_TOOLCHAIN_VARIANT=arcmwdt

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: board: nsim: enable metaware toolchain in sanitycheck

Enable MWDT toolchain in sanitycheck for all ARC nSIM boards

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
isr_tables: don't whole-archive library

As of today we have a bit weird situation with generated
sw_isr_table / irq_vector_table tables.

On the final linkage stage we pass two files which content
section with sw_isr_table / irq_vector_table. They are
 * libarch__common.a (with an outdated tables from the first
   linkage stage)
 * isr_tables.c.obj (with an actual tables)

The sections where tables are located are marked with
".gnu.linkonce" prefix. That means:
<<<As a GNU extension, if the name begins with .gnu.linkonce,
we only link a single copy of the section.>>>

However the "libarch__common.a" is passed to linker with
"--whole-archive" option which means <<<include every object
file in the archive in the link, rather than searching the archive
for the required object files>>>

That combination confuses MWDT linker and breaks linkage with
MWDT toolchain.

As a simple fix we can move the sw_isr_table / irq_vector_table
sections to their own library and link this library with
"--no-whole-archive" option.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: sys_io: rewrite to C code

Replace ASM sys_io implementation with identical C code for ARC.
This significantly improves portability, i.e. compiler decides
which instructions to use for a particular CPU and / or
configuration.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: use MWDT intrinsics to access aux regs in case of MWDT toolchain

Metaware doesn't support gcc's builtins so use corresponding intrinsics
instead.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
app_memdomain: ARC: adjust asembler directives for MWDT toolchain

ARC MWDT assembler has slightly different directives names:
.pushsect instead of .pushsection
.popsect instead of .popsection

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
linker: tweak section naming to feet all linkers

MWDT toolchain adds additional suffix to sections name in case of
ffunction-sections / fdata-sections are enabled.

As proposed by Andy Ross let's pick a single set of rules
and syntax that work.

Suggested-by: Andy Ross <andy@plausible.org>
Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: linker: implement linker related definition for MWDT

Implement linker related definitions and macros for MWDT
toolchain.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: toolchain: handle difference in macros for MWDT and ARC GNU

Create include/toolchain/mwdt.h to handle difference in
several macros for MWDT and ARC GNU toolchains.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: soc: add MWDT compiler options for nsim targets

Add metaware compiler options for nsim targets

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
cmake: allow to specify compliler for DTS preprocessing

MWDT CCAC compiler can't handle the dts file as gcc does, so
allow to specify compliler for DTS preprocessing which is differ
from ${CMAKE_C_COMPILER}.

Use ARC GNU compiler for DTS preprocessing in case of MWDT
toolchain.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

evgeniy-paltsev commented:
@tejlmand @nashif @andyross @SebastianBoe 

I've pushed reworked series. It's addressed the previous comments and includes significant code cleanup.
Could you please review it? It would be nice to sneak in v2.4 release.


I've run sanitycheck for 
qemu_cortex_a53, qemu_cortex_m0, qemu_cortex_m3, qemu_cortex_r5, qemu_nios2, qemu_riscv32, qemu_riscv64, qemu_x86, qemu_xtensa, nsim_em, nsim_hs, nsim_sem, nsim_sem_mpu_stack_guard
platforms - no failing tests.

abrodkin reviewed (approved):
With that most recent re-spin it looks pretty clean and "almost harmless" (C) for the rest of architectures except ARC.

ruuddw commented on include/toolchain/mwdt.h:
```diff
@@ -0,0 +1,79 @@
+/* Metaware toolchain has _Static_assert. However it not able to calculate
+ * conditional expression in build time for some realy complex cases. ARC GNU
+ * toolchain works fine in this cases, so it looks like MWDT bug. So, disable
+ * BUILD_ASSERT macro until we fix that issue in MWDT toolchain.
```
Please make sure to create an issue for this so we don't forget.

ruuddw commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -7,6 +10,34 @@ zephyr_compile_options_ifdef(CONFIG_FPU -mfpu=fpud_all)
+else()
+# MWDT compiler options
+zephyr_compile_options_ifdef(CONFIG_SOC_NSIM_EM -arcv2em -core3 -Xdiv_rem=radix2
+			     -Xmpy_option=mpyd -Xbitscan -Xswap -Xbarrel_shifter
```
With MetaWare, we could use a tcf file instead of these.

evgeniy-paltsev replied:
Moreover the .tcf file may includes the properties for nSIM and MDB as well. So, I propose to have it as a separate enhancement - to switch all metaware tools (compiler, simulator, debugger) to use tcf.

ruuddw replied:
Agreed, this is a separate topic and should extend to other tools (maybe even including ARC gnu).

ruuddw reviewed (approved):
Great work, I have a few minor comments but these can be later improvements as well (given the merge window deadline)

evgeniy-paltsev commented on include/arch/arc/v2/linker-mwdt.ld:
```diff
@@ -0,0 +1,283 @@
+		*(".text.*")
+
+/* mwdt version of kobject-text.ld
+ * FixMe: mwdt's linker does not support the following statement:
```
Ooops, I've forgotten to drop this comment, my fault.

evgeniy-paltsev commented on cmake/compiler/arcmwdt/generic.cmake:
```diff
@@ -0,0 +1,28 @@
+
+# Configures CMake for using ccac
+
+find_program(CMAKE_DTS_PREPROCESS_C_COMPILER arc-elf32-gcc)
```
CCAC can't be used here as it can't handle DTS files (that's why we create this workaround for metaware toolchain)
I can replace ARC GCC usage with host GCC (however I'd prefer to use ARC one)

abrodkin replied:
@ruuddw see internal Jira issue `P10019563-38578` for all the gory details.

ruuddw replied:
If the ccac compiler driver is 'too clevel', just use clangac directly?
But, this can be changed later, when the objcopy dependency is resolved as well.

tejlmand replied:
Be careful here.
Could be wise to check for both host gcc, as that will most likely be available on Linux machines, however, on Windows, then it might be more likely to find a cross gcc variant.

So probably check for multiple options, and use the best match.
(See also other comment regarding `objcopy` for inspiration.)

tejlmand replied:
> CCAC can't be used here as it can't handle DTS files 

Could you elaborate on what issues are seen ?
Maybe @mbolivar-nordic  or @galak would be able to evaluate if DTS files could be made in a more non-GCC friendly way.

evgeniy-paltsev replied:
The main issue is quite simple - MWDT compiler has strict restrictions about file extensions, so it throws error when gets something.dts

evgeniy-paltsev commented on arch/arc/core/fault_s.S:
```diff
@@ -104,7 +104,7 @@ _exc_entry:
 	mov_s r0, sp
 	/* ilink is the thread's original sp */
-	mov_s r1, ilink
+	mov r1, ilink
```
No, ARC GCC chooses short or long instruction encoding itself :)

ilink can't be accessed with mov_s instruction, however ARC GNU assembler gracefully allows us to use mov_s here and implicitly replace it with mov.

MWDT assembler thinks that it is outrageous and generate error here.

abrodkin replied:
@ruuddw let's put it like that: MW compiler here is doing much a stricter check against the PRM, even "stricter" than real HW :)

IRISZZW commented:
When run sanitycheck with arcmwdt toolchain, you need using [-W option](https://github.com/zephyrproject-rtos/zephyr/blob/4eba946e2bb1fe97b2a9ac65c4c85400f87e226a/scripts/sanitycheck#L416-L417) to disable erroring on warnings
https://github.com/zephyrproject-rtos/zephyr/blob/4eba946e2bb1fe97b2a9ac65c4c85400f87e226a/scripts/sanity_chk/sanitylib.py#L1833-L1841

tejlmand commented on arch/common/CMakeLists.txt:
```diff
@@ -39,3 +38,29 @@ if (DEFINED CONFIG_ARM OR DEFINED CONFIG_X86
+  add_dependencies(isr_tables zephyr_generated_headers)
+  target_link_libraries(isr_tables zephyr_interface)
+
+  set_property(GLOBAL APPEND PROPERTY ZEPHYR_NO_WHOLE_ARCHIVE_LIBS isr_tables)
```
This should not be needed.
When using a normal CMake library that is not `--whole-archive`'d, then we should be using normal CMake library dependencies.

That would be, the libraries that is dependent upon `isr_tables` should have a line, similar to:
```
zephyr_library_link_libraries_ifdef(CONFIG_GEN_ISR_TABLES isr_tables)
```

As example, this would work for arm targets:
```
diff --git a/arch/arm/core/aarch32/CMakeLists.txt b/arch/arm/core/aarch32/CMakeLists.txt
index 4d2898970e..d419b5fbf3 100644
--- a/arch/arm/core/aarch32/CMakeLists.txt
+++ b/arch/arm/core/aarch32/CMakeLists.txt
@@ -34,3 +34,4 @@ add_subdirectory_ifdef(CONFIG_ARM_NONSECURE_FIRMWARE cortex_m/tz)
 add_subdirectory_ifdef(CONFIG_CPU_CORTEX_R cortex_a_r)
 
 zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors vector_table.ld)
+zephyr_library_link_libraries_ifdef(CONFIG_GEN_ISR_TABLES isr_tables)
```

and thus remove the need for the new `ZEPHYR_NO_WHOLE_ARCHIVE_LIBS` property.

tejlmand commented on cmake/app/boilerplate.cmake:
```diff
@@ -42,6 +42,12 @@ define_property(GLOBAL PROPERTY ZEPHYR_LIBS
 zephyr_library() appends libs to this list.")
 set_property(GLOBAL PROPERTY ZEPHYR_LIBS "")
 
+define_property(GLOBAL PROPERTY ZEPHYR_NO_WHOLE_ARCHIVE_LIBS
+    BRIEF_DOCS "Global list of all CMake libs that shouldn't be whole-archive'd"
+    FULL_DOCS  "Global list of all CMake libs that should be linked in but
+should not be --whole-archive'd.")
+set_property(GLOBAL PROPERTY ZEPHYR_NO_WHOLE_ARCHIVE_LIBS "")
+
```
Not `whole-archiving` is default behavior in CMake, so we should not need to have special handling of that.

ruuddw commented:
Still some coding style compliance check issues to be fixed, for the rest I'm ok.

tejlmand commented on cmake/bintools/arcmwdt/target.cmake:
```diff
@@ -0,0 +1,356 @@
+      endif()
+    endforeach()
+    set(obj_copy_cmd
+        COMMAND arc-elf32-objcopy ${obj_copy_sections_rename}
```
> Could host objcopy be used instead?

That should most certainly work on Linux / MacOS, but might be a little trickier on Windows, cause there are lot of Windows developers who doesn't install a native GNU Toolchain/bintools.

>Maybe instead of hardcoding the tool here, add a function to find it/resolve the path and warn if not found?

agree, no hard coding, cause that may lead to failure during build time with a cryptic message, instead of CMake configure phase.
Also a much better error message can be provided to the user.
So something like:
```
find_program(CMAKE_OBJCOPY objcopy)
find_program(CMAKE_ARC_OBJCOPY arc-elf32-objcopy)
if(CMAKE_OBJCOPY)
   ... do stuff
elseif(CMAKE_ARC_OBJCOPY)
   ... do other stuff
else()
  message(FATAL_ERROR "A good error message")
endif()
```

tejlmand commented on cmake/linker/ld/target.cmake:
```diff
@@ -92,7 +92,7 @@ function(toolchain_ld_link_elf)
     ${ZEPHYR_LIBS_PROPERTY}
     ${LINKERFLAGPREFIX},--no-whole-archive
-    kernel
+    ${ZEPHYR_NO_WHOLE_ARCHIVE_LIBS_PROPERTY}
```
please keep original implementation.
```suggestion
    kernel
```

tejlmand commented on cmake/toolchain/arcmwdt/generic.cmake:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+set_ifndef(MWDT_TOOLCHAIN_PATH "$ENV{MWDT_TOOLCHAIN_PATH}")
```
I would suggest that because the toolchain variant is named: `arcmwdt`, then use the name: `ARCMWDT_TOOLCHAIN_PATH`

That would be aligned with:
`gnuarmemb`: `GNUARMEMB_TOOLCHAIN_PATH`
`espressif`: `ESPRESSIF_TOOLCHAIN_PATH`
`xtools`: `XTOOLS_TOOLCHAIN_PATH`

evgeniy-paltsev replied:
Nice catch, fixed

tejlmand commented on kernel/CMakeLists.txt:
```diff
@@ -69,3 +69,5 @@ target_include_directories(kernel PRIVATE
 add_dependencies(kernel zephyr_generated_headers)
 
 target_link_libraries(kernel zephyr_interface)
+
+set_property(GLOBAL APPEND PROPERTY ZEPHYR_NO_WHOLE_ARCHIVE_LIBS kernel)
```
Can be dropped. See other comments.
```suggestion
```

tejlmand commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -7,6 +10,34 @@ zephyr_compile_options_ifdef(CONFIG_FPU -mfpu=fpud_all)
 else()
 zephyr_compile_options_ifdef(CONFIG_FPU -mfpu=fpuda_all)
 endif()
+else()
```
I know you just follow existing style, but I think indentation should be fixed.
But it is accepted if done at later time, that is after this PR.

```
if()
...
...
if()
..
else()
..
endif()
else()
...
...
```
is hard to read.

tejlmand reviewed (changes requested):
First round of build system related code reviewed.
Did some verification of Zephyr SDK toolchain still works.
Need to follow up regarding MWDT.

tejlmand commented on cmake/dts.cmake:
```diff
@@ -122,6 +122,10 @@ if(SUPPORTS_DTS)
   set(CACHED_DTS_ROOT_BINDINGS ${DTS_ROOT_BINDINGS} CACHE INTERNAL
     "DT bindings root directories")
 
+  if(NOT DEFINED CMAKE_DTS_PREPROCESS_C_COMPILER)
+    set(CMAKE_DTS_PREPROCESS_C_COMPILER ${CMAKE_C_COMPILER})
```
CMake uses the form: `CMAKE_<LANG>_COMPILER`
https://cmake.org/cmake/help/v3.0/variable/CMAKE_LANG_COMPILER.html

So let's just call it.
```suggestion
  if(NOT DEFINED CMAKE_DTS_COMPILER)
    set(CMAKE_DTS_COMPILER ${CMAKE_C_COMPILER})
```

tejlmand commented on cmake/dts.cmake:
```diff
@@ -133,7 +137,7 @@ if(SUPPORTS_DTS)
   # so that changes to DT sources are detected.
   execute_process(
-    COMMAND ${CMAKE_C_COMPILER}
+    COMMAND ${CMAKE_DTS_PREPROCESS_C_COMPILER}
```
```suggestion
    COMMAND ${CMAKE_DTS_COMPILER}
```

tejlmand commented on include/arch/arc/v2/linker-mwdt.ld:
```diff
@@ -0,0 +1,283 @@
+/*
```
It seems that `linker-gnu.ld` and `linker-mwdt.ld` does have much in common.
We should try to see if they could be aligned, but it can go in for now.

evgeniy-paltsev replied:
Yep, I'd better keep it as a separate enhancement.

tejlmand replied:
agreed. it's for later improvement

tejlmand commented on include/linker/sections.h:
```diff
@@ -31,8 +31,8 @@
-#define _IRQ_VECTOR_TABLE_SECTION_NAME	.gnu.linkonce.irq_vector_table
-#define _SW_ISR_TABLE_SECTION_NAME	.gnu.linkonce.sw_isr_table
+#define _IRQ_VECTOR_TABLE_SECTION_NAME	.gnu.linkonce.irq_vector_table*
+#define _SW_ISR_TABLE_SECTION_NAME	.gnu.linkonce.sw_isr_table*
```
There was a question from @sebo:
> Do we know why mwdt is generating the . suffix and gcc is not?

could you answer that ?

abrodkin replied:
@tejlmand , @SebastianBoe 

> There was a question from @sebo:
> 
> > Do we know why mwdt is generating the . suffix and gcc is not?
> 
> could you answer that ?

@tejlmand, @SebastianBoe as @evgeniy-paltsev mentioned in the commit message here https://github.com/zephyrproject-rtos/zephyr/pull/22668/commits/8169f2d586ba973145a7bbb8789eda44628e9e86 that's a behavior of MetaWare toolchain and it differs from GCC. I'm not a compiler expert and not sure if there's an established standard for section names generation. If there's such a standard I'll happily point to it our MetaWare developers with request to accommodate it.

tejlmand replied:
thanks

tejlmand commented on include/arch/arc/v2/linker-mwdt.ld:
```diff
@@ -0,0 +1,283 @@
+ * zephyr_linker_sources() Cmake function.
+ */
+#include <snippets-rom-start.ld>
+		*(".text")
+		*(".text*")
+		*(".text.*")
```
@sebo had a suggestion here:
> It's cleaner to do
> 
> `*(.text .text* .text.*)`
> Then the linker is not forced to place these input sections after each other and has more freedom to optimize.

evgeniy-paltsev replied:
done

tejlmand commented on cmake/compiler/arcmwdt/generic.cmake:
```diff
@@ -2,6 +2,12 @@
 
+find_program(CMAKE_DTS_PREPROCESS_C_COMPILER arc-elf32-gcc)
+
+if(CMAKE_DTS_PREPROCESS_C_COMPILER STREQUAL CMAKE_DTS_PREPROCESS_C_COMPILER-NOTFOUND)
```
Can be simplified:
```suggestion
if(CMAKE_DTS_COMPILER)
```

> if(<constant>)
>
>    False if the constant is 0, OFF,..., or ends in the suffix -NOTFOUND.

evgeniy-paltsev replied:
Ok, I'll fix the check, however I'll use CMAKE_DTS_PREPROCESS_C_COMPILER variable name as we already have 'dtc' which is dts compiler (so CMAKE_DTS_COMPILER name will be confusing)

tejlmand replied:
accepted, `CMAKE_DTS_PREPROCESSOR` is fine.

evgeniy-paltsev replied:
done

tejlmand reviewed (changes requested):
Some more, minor cleanups.

tejlmand committed:
cmake: mwdt compiler adopted to new toolchain abstraction

This commit converts the MWDT compiler implementation from the old
macro based approach and into the new toolchain property scheme.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

Commit:
ARC: SOC: NSIM: fix CMakeLists indentation

No functional changes intended

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: linker: add more place for optimization

Do not force linker to place text sections after each other
to have more freedom to optimize.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

ruuddw commented:
@galak : why shouldn't this go into 2.4? Torsten and Eugeniy addressed all relevant comments, and 3rd party toolchain support is a goal for this release.

evgeniy-paltsev commented:
I've pushed new changes related on the previous feedback:
 * drop all ZEPHYR_NO_WHOLE_ARCHIVE_LIBS_PROPERTY mess
 * don't hardcode arc-elf32-objcopy but search for available objcopy (ARC or host one)
 * allow to use host GCC for DTS preprocessing, not only ARC one
 * rename CMAKE_DTS_PREPROCESS_C_COMPILER to CMAKE_DTS_PREPROCESSOR
 * use ARCMWDT_TOOLCHAIN_PATH instead of MWDT_TOOLCHAIN_PATH environment variable
 * lot's of style-related cleanups

tejlmand commented on cmake/bintools/arcmwdt/target.cmake:
```diff
@@ -0,0 +1,371 @@
+
+  # Handle stripping
+  if (DEFINED BINTOOLS_OBJCOPY_STRIP OR DEFINED BINTOOLS_OBJCOPY_STRIP_ALL)
+    if(${BINTOOLS_OBJCOPY_STRIP_ALL})
+      set(obj_copy_strip "-qs")
+    elseif(${BINTOOLS_OBJCOPY_STRIP_DEBUG})
+      set(obj_copy_strip "-ql")
+    endif()
+    set(obj_copy_cmd
+        COMMAND ${CMAKE_STRIP} ${obj_copy_strip}
+        ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
+  endif()
+
+  # no support of --srec-len in mwdt
+
+  # Handle Input and Output target types
+  if(DEFINED BINTOOLS_OBJCOPY_TARGET_OUTPUT)
+    set(obj_copy_target_output "")
+    set(obj_copy_gap_fill "")
+    if(DEFINED BINTOOLS_OBJCOPY_GAP_FILL)
+    set(obj_copy_gap_fill "-f;${BINTOOLS_OBJCOPY_GAP_FILL}")
+    endif()
+      # only mwdt's elf2hex supports gap fill
+    if(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "srec")
+      set(obj_copy_target_output "-m")
+    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "ihex")
+      set(obj_copy_target_output "-I")
+    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "binary")
+      set(obj_copy_target_output "-B")
+    endif()
+     set(obj_copy_cmd
+        COMMAND ${CMAKE_ELF2HEX} ${obj_copy_gap_fill} ${obj_copy_target_output}
+        -o ${BINTOOLS_OBJCOPY_FILE_OUTPUT} ${BINTOOLS_OBJCOPY_FILE_INPUT})
```
This will not work as expected.
When this macro is being called from here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L1134-L1142
then it is called with `STRIP_ALL` and `TARGET_OUTPUT ihex`, which will result in first, line 132-134:
```
   set(obj_copy_cmd
       COMMAND ${CMAKE_STRIP} ${obj_copy_strip}
        ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
```
which will take the `FILE_INPUT` and create `FILE_OUTPUT`.
But then at line 154-156, this happens:
```
     set(obj_copy_cmd
        COMMAND ${CMAKE_ELF2HEX} ${obj_copy_gap_fill} ${obj_copy_target_output}
        -o ${BINTOOLS_OBJCOPY_FILE_OUTPUT} ${BINTOOLS_OBJCOPY_FILE_INPUT})
```
which basically means that the original `FILE_INPUT` is used again and then overwriting `FILE_OUTPUT` which effectively will result in the `STRIP_ALL` to have no effect.

ruuddw replied:
Doesn't elf2hex strip out any debug info anyway, leaving only the binary code and no debug sections?

abrodkin replied:
Good catch! So that basically happens because in existing case with GNU toolchain we do both manipulations in 1 go here https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-mwdt/cmake/bintools/gnu/target_objcopy.cmake#L111:
```cmake
  # Construct the command
  set(obj_copy_cmd
    # Base command
    COMMAND ${CMAKE_OBJCOPY} ${obj_copy_strip} ${obj_copy_gap_fill} ${obj_copy_srec_len}
    # Input and Output target types
    ${obj_copy_target_input} ${obj_copy_target_output}
    # Sections
    ${obj_copy_sections_only} ${obj_copy_sections_rename} ${obj_copy_sections_remove}
    # Input and output files
    ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT}
  )
```
while in case of MW tools we don't have a direct substitution of really powerful `objcopy` and have a multi-stage pipeline of binary processing, thus separate calls to `stripac`, `elf2hex` etc.

Moreover [though I'm far from being even an experienced user of CMake] I'm not sure if all those manipulations are being done since we do `set(obj_copy_cmd ...` multiple times... I'd think that way we **re-define** the same command again and again. Correct me if I'm wrong though.
```
  # Handle Input and Output target types
  if(DEFINED BINTOOLS_OBJCOPY_TARGET_OUTPUT)
    set(obj_copy_target_output "")
    set(obj_copy_gap_fill "")
    if(DEFINED BINTOOLS_OBJCOPY_GAP_FILL)
    set(obj_copy_gap_fill "-f;${BINTOOLS_OBJCOPY_GAP_FILL}")
    endif()
      # only mwdt's elf2hex supports gap fill
    if(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "srec")
      set(obj_copy_target_output "-m")
    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "ihex")
      set(obj_copy_target_output "-I")
    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "binary")
      set(obj_copy_target_output "-B")
    endif()
     set(obj_copy_cmd                                           <-- *** HERE ***
        COMMAND ${CMAKE_ELF2HEX} ${obj_copy_gap_fill} ${obj_copy_target_output}
        -o ${BINTOOLS_OBJCOPY_FILE_OUTPUT} ${BINTOOLS_OBJCOPY_FILE_INPUT})
  endif()

  # Handle sections, if any
  # 1. Section only selection(s)
  set(obj_copy_sections_only "")
  if(DEFINED BINTOOLS_OBJCOPY_SECTION_ONLY)
    foreach(section_only ${BINTOOLS_OBJCOPY_SECTION_ONLY})
      list(APPEND obj_copy_sections_only "-sn;${section_only}")
    endforeach()
    set(obj_copy_cmd                                           <-- *** HERE ***
        COMMAND ${CMAKE_ELF2BIN} -q ${obj_copy_sections_only}
        ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
  endif()
```

Still given we don't currently have any ARC boards supported by upstream Zephyr which are programmed with binaries (we load Elf contents with either OpenOCD or MDB via JTAG) this doesn't affect `sanitychecker` results and given current race to the feature freeze I guess we may address it via formal bug-fix during stabilization period. Agree?

tejlmand replied:
> (we load Elf contents with either OpenOCD or MDB via JTAG)

agreed, then we can take it in, as-is.

> set(obj_copy_cmd ...

right, the last `set` wins.
(it should have been `list(APPEND ...)` in order to work.)

tejlmand reviewed (changes requested):
Found an error in bintools handling

tejlmand committed:
cmake: mwdt bintools adopted to new toolchain abstraction

This commit converts the MWDT bintools implementation from the old
macro based approach and into the new toolchain property scheme.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
rebased on top of master, and with two additional commits for adopting to the new toolchain abstraction:
https://github.com/tejlmand/zephyr/commits/prototyping/16031_toolchain_abstraction_mwdt

tejlmand commented on cmake/bintools/arcmwdt/target.cmake:
```diff
@@ -0,0 +1,371 @@
+  endif()
+
+  # Handle stripping
+  if (DEFINED BINTOOLS_OBJCOPY_STRIP OR DEFINED BINTOOLS_OBJCOPY_STRIP_ALL)
```
This is fixed here: https://github.com/tejlmand/zephyr/commits/prototyping/16031_toolchain_abstraction_mwdt

tejlmand reviewed (approved):
Approved. Just needs a rebase and the new toolchain abstraction commits.

tejlmand commented:
@andyross please revisit

tejlmand commented:
@daor-oti please revisit

tejlmand commented:
> > rebased on top of master, and with two additional commits for adopting to the new toolchain abstraction:
> > https://github.com/tejlmand/zephyr/commits/prototyping/16031_toolchain_abstraction_mwdt
> 
> @tejlmand there seem to be a missing ")" in your branch:
> 

Fixed, it was two leftover lines from the adoptation, line 25-26:
```
macro(toolchain_cc_warning_base)
  zephyr_compile_options(
```
is now removed.

abrodkin commented:
> > > rebased on top of master, and with two additional commits for adopting to the new toolchain abstraction:
> > > https://github.com/tejlmand/zephyr/commits/prototyping/16031_toolchain_abstraction_mwdt
> > 
> > 
> > @tejlmand there seem to be a missing ")" in your branch:
> 
> Fixed, it was two leftover lines from the adoptation, line 25-26:
> 
> ```
> macro(toolchain_cc_warning_base)
>   zephyr_compile_options(
> ```
> 
> is now removed.

That I figured out myself just before your update, though thanks for the quick fix :)
Now more things.
1. We used to compile with `-std=gnu99` and it was OKish, now with default being set for all to `-std=c99` and no override for MWDT I start to see unmutable by `-Wno-main-return-type` error:
   ```
   .../zephyrproject/zephyr/samples/hello_world/src/main.c:10:1: error: 'main' must 
   return 'int'
   void main(void)
   ^~~~
   int
   ```
   with the next hunk in place this one is gone:
   ```diff
   diff --git a/cmake/compiler/arcmwdt/compiler_flags.cmake b/cmake/compiler/arcmwdt/compiler_flags.cmake
   index 77aeeb5510..44847ddf14 100644
   --- a/cmake/compiler/arcmwdt/compiler_flags.cmake
   +++ b/cmake/compiler/arcmwdt/compiler_flags.cmake
   @@ -170,6 +170,8 @@ set_compiler_property(PROPERTY security_fortify "")
    # Required C++ flags when using mwdt
    set_property(TARGET compiler-cpp PROPERTY required "")
   
   +set_property(GLOBAL PROPERTY CSTD gnu99)
   +
    #################################
    # This section covers asm flags #
    #################################
   ```
2. But then the following happens (obviously on clean re-build):
   ```
   [ 58%] Generating linker.cmd
   cd .../zephyrproject/zephyr/build/zephyr && /global/apps/mwdt_2020.06/MetaWare/arc/bin/ccac -x c -Hnocopyr -MD -MF linker.cmd.dep -MT zephyr/linker.cmd -D_LINKER -D_ASMLANGUAGE "-I.../zephyrproject/zephyr/include -I.../zephyrproject/zephyr/build/zephyr/include/generated -I.../zephyrproject/zephyr/soc/arc/snps_nsim -I.../zephyrproject/zephyr/lib/libc/minimal/include" -D__MWDT_LINKER_CMD__ .../zephyrproject/zephyr/soc/arc/snps_nsim/linker.ld -P -o linker.cmd
   .../zephyrproject/zephyr/soc/arc/snps_nsim/linker.ld:11:10: fatal error: 'devicetree.h' file not found
    #include <devicetree.h>
            ^~~~~~~~~~~~~~
   ```

abrodkin commented:
> > > > rebased on top of master, and with two additional commits for adopting to the new toolchain abstraction:
> > > > https://github.com/tejlmand/zephyr/commits/prototyping/16031_toolchain_abstraction_mwdt
> > > 
> > > 
> > > @tejlmand there seem to be a missing ")" in your branch:
> > 
> > 
> > Fixed, it was two leftover lines from the adoptation, line 25-26:
> > ```
> > macro(toolchain_cc_warning_base)
> >   zephyr_compile_options(
> > ```
> > 
> > 
> > is now removed.
> 
> That I figured out myself just before your update, though thanks for the quick fix :)
> Now more things.
> 
> 1. We used to compile with `-std=gnu99` and it was OKish, now with default being set for all to `-std=c99` and no override for MWDT I start to see unmutable by `-Wno-main-return-type` error:
>    ```
>    .../zephyrproject/zephyr/samples/hello_world/src/main.c:10:1: error: 'main' must 
>    return 'int'
>    void main(void)
>    ^~~~
>    int
>    ```
>    
>    
>    with the next hunk in place this one is gone:
>    ```diff
>    diff --git a/cmake/compiler/arcmwdt/compiler_flags.cmake b/cmake/compiler/arcmwdt/compiler_flags.cmake
>    index 77aeeb5510..44847ddf14 100644
>    --- a/cmake/compiler/arcmwdt/compiler_flags.cmake
>    +++ b/cmake/compiler/arcmwdt/compiler_flags.cmake
>    @@ -170,6 +170,8 @@ set_compiler_property(PROPERTY security_fortify "")
>     # Required C++ flags when using mwdt
>     set_property(TARGET compiler-cpp PROPERTY required "")
>    
>    +set_property(GLOBAL PROPERTY CSTD gnu99)
>    +
>     #################################
>     # This section covers asm flags #
>     #################################
>    ```
> 2. But then the following happens (obviously on clean re-build):
>    ```
>    [ 58%] Generating linker.cmd
>    cd .../zephyrproject/zephyr/build/zephyr && /global/apps/mwdt_2020.06/MetaWare/arc/bin/ccac -x c -Hnocopyr -MD -MF linker.cmd.dep -MT zephyr/linker.cmd -D_LINKER -D_ASMLANGUAGE "-I.../zephyrproject/zephyr/include -I.../zephyrproject/zephyr/build/zephyr/include/generated -I.../zephyrproject/zephyr/soc/arc/snps_nsim -I.../zephyrproject/zephyr/lib/libc/minimal/include" -D__MWDT_LINKER_CMD__ .../zephyrproject/zephyr/soc/arc/snps_nsim/linker.ld -P -o linker.cmd
>    .../zephyrproject/zephyr/soc/arc/snps_nsim/linker.ld:11:10: fatal error: 'devicetree.h' file not found
>     #include <devicetree.h>
>             ^~~~~~~~~~~~~~
>    ```

@tejlmand OK, actually (2) is there anyways, we just not getting that far in case of `-std=c99` stopping build due to an error much earlier.
Let me see how that used to work before.

tejlmand commented:
The default is `C99` if nothing is defined:
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L894-L895
```
get_property(CSTD GLOBAL PROPERTY CSTD)
set_ifndef(CSTD c99)
```

abrodkin commented:
> The default is `C99` if nothing is defined:
> https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L894-L895
> 
> ```
> get_property(CSTD GLOBAL PROPERTY CSTD)
> set_ifndef(CSTD c99)
> ```

That I understood as well (see my fix above), now trying to find which include path is missing so that `devicetree.h` couldn't be found. Stay tuned :)

tejlmand commented:
> That way no include paths are being understood I guess. Know where that's done?

that is probably a space separated list, which is not a real list.
Let me try to see if I can spot the reason.

tejlmand commented:
> > That way no include paths are being understood I guess. Know where that's done?
> 
> that is probably a space separated list, which is not a real list.
> Let me try to see if I can spot the reason.

Apply this patch:
```
diff --git a/cmake/linker/arcmwdt/target.cmake b/cmake/linker/arcmwdt/target.cmake
index d864be148b..2f3f0e2683 100644
--- a/cmake/linker/arcmwdt/target.cmake
+++ b/cmake/linker/arcmwdt/target.cmake
@@ -26,7 +26,7 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
     set(linker_script_dep "")
   endif()
 
-  zephyr_get_include_directories_for_lang(C current_includes)
+  zephyr_get_include_directories_for_lang(C current_includes "$<SEMICOLON>")
   get_filename_component(base_name ${CMAKE_CURRENT_BINARY_DIR} NAME)
   get_property(current_defines GLOBAL PROPERTY PROPERTY_LINKER_SCRIPT_DEFINES)
```

tejlmand committed:
cmake: MWDT linker command list fix.

This commit fixes an issue where the list returned by
zephyr_get_include_directories_for_lang() would be a space separated
list, cause the arguments to be quoted during linking.

This is fixed by requesting `$<SEMICOLON>` to be used as separated which
which again ensures a correct expansion of the list.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
branch updated as well.

abrodkin commented:
> diff --git a/cmake/linker/arcmwdt/target.cmake b/cmake/linker/arcmwdt/target.cmake
> index d864be148b..97667a9ed9 100644
> --- a/cmake/linker/arcmwdt/target.cmake
> +++ b/cmake/linker/arcmwdt/target.cmake
> @@ -26,7 +26,7 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
>      set(linker_script_dep "")
>    endif()
>  
> -  zephyr_get_include_directories_for_lang(C current_includes)
> +  zephyr_get_include_directories_for_lang(C current_includes "$<SEMICOLON>")
>    get_filename_component(base_name ${CMAKE_CURRENT_BINARY_DIR} NAME)
>    get_property(current_defines GLOBAL PROPERTY PROPERTY_LINKER_SCRIPT_DEFINES)
>  
> @@ -53,7 +53,8 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
>      -o ${linker_script_gen}
>      VERBATIM
>      WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
> -  )
> +    COMMAND_EXPAND_LISTS
> +)
>  endmacro()
>  

Cool, that one is gone now (no more extra `"` so we pass through):
```
[ 13%] Generating linker.cmd
cd .../zephyrproject/zephyr/build/zephyr && /global/apps/mwdt_2020.06/MetaWare/arc/bin/ccac -x c -Hnocopyr -MD -MF linker.cmd.dep -MT zephyr/linker.cmd -D_LINKER -D_ASMLANGUAGE -I.../zephyrproject/zephyr/include -I.../zephyrproject/zephyr/build/zephyr/include/generated -I.../zephyrproject/zephyr/soc/arc/snps_nsim -I.../zephyrproject/zephyr/lib/libc/minimal/include -D__MWDT_LINKER_CMD__ .../zephyrproject/zephyr/soc/arc/snps_nsim/linker.ld -P -o linker.cmd
```

Now the next one:
```
[ 63%] Building ASM object zephyr/arch/arch/arc/core/CMakeFiles/arch__arc__core.dir/thread_entry_wrapper.S.obj
cd .../zephyrproject/zephyr/build/zephyr/arch/arch/arc/core && /global/apps/mwdt_2020.06/MetaWare/arc/bin/ccac -DBUILD_VERSION=zephyr-v2.3.0-2685-g1aa8db1fee9b -DKERNEL -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I.../zephyrproject/zephyr/kernel/include -I.../zephyrproject/zephyr/arch/arc/include -I.../zephyrproject/zephyr/include -I.../zephyrproject/zephyr/build/zephyr/include/generated -I.../zephyrproject/zephyr/soc/arc/snps_nsim -I.../zephyrproject/zephyr/lib/libc/minimal/include -I/global/apps/mwdt_2020.06/MetaWare/arc/inc  -Os -imacros .../zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -Hnocrt -fno-common -g -Hnocopyr -HL -Hnosdata -Hasmcpp -imacros .../zephyrproject/zephyr/include/toolchain/zephyr_stdint.h -D_ASMLANGUAGE -Wno-sometimes-uninitialized -Wno-shift-overflow -Wno-missing-braces -Wno-self-assign -Wno-address-of-packed-member -Wno-unused-function -Wno-initializer-overrides -Wno-section -Wno-unknown-warning-option -Wno-unused-variable -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -ffunction-sections -fdata-sections -arcv2em -core3 -Xdiv_rem=radix2 -Xmpy_option=mpyd -Xbitscan -Xswap -Xbarrel_shifter -Xshift_assist -Xdsp2 -Xdsp_complex -Xdsp_divsqrt=radix2 -Xdsp_itu -Xdsp_accshift=full -Xfpus_div -Xfpu_mac -Xfpuda -Xfpus_mpy_slow -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1 -Hno_default_include -Hnoarcexlib -o CMakeFiles/arch__arc__core.dir/thread_entry_wrapper.S.obj -c .../zephyrproject/zephyr/arch/arc/core/thread_entry_wrapper.S
[/global/apps/mwdt_2020.06/MetaWare/arc/bin/asac] ".../zephyrproject/zephyr/arch/arc/core/thread_entry_wrapper.S",Line 30: error: syntax error
```
Looking into it.

tejlmand commented:
> > branch updated as well.
> 
> What about this fix which we **do** need?:
> 
> ```
> set_property(GLOBAL PROPERTY CSTD gnu99)
> ```

That is usually defined by the board:
https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/posix/native_posix/CMakeLists.txt#L43
https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/posix/nrf52_bsim/CMakeLists.txt#L57

but for now, I will say it's acceptable you set this in MWDT toolchain file.

tejlmand commented:
> 

> What about this fix which we **do** need?:
> 
> ```
> set_property(GLOBAL PROPERTY CSTD gnu99)
> ```

fix added to branch and forced pushed.

nashif commented on include/arch/arc/v2/linker.ld:
```diff
@@ -1,221 +1,12 @@
 
-#if defined(CONFIG_XIP)
-	#define _DATA_IN_ROM __data_rom_start
+#if defined(__MWDT_LINKER_CMD__)
```
this is going to be a recipe of issues, where someone changes for one and does not update the other because we do not have mwdt in CI... There needs to be one place for linker data, but I guess we have no way around this right now.

tejlmand replied:
@nashif  agree, and also commented here:
https://github.com/zephyrproject-rtos/zephyr/pull/22668#discussion_r483579164

> It seems that linker-gnu.ld and linker-mwdt.ld does have much in common.
> We should try to see if they could be aligned, but it can go in for now.

answer:
> Yep, I'd better keep it as a separate enhancement.

abrodkin commented:
@tejlmand OK so yet another problem is gone with this:
```diff
diff --git a/include/toolchain/mwdt.h b/include/toolchain/mwdt.h
index f7139738a1..a8bcc97c21 100644
--- a/include/toolchain/mwdt.h
+++ b/include/toolchain/mwdt.h
@@ -17,12 +17,12 @@
 #define FUNC_INSTR(a)

 .macro section_var_mwdt, section, symbol
-       .section .\&section\&.\ &symbol, "aw"
+       .section .\&section\&.\&symbol, "aw"
        symbol :
 .endm

 .macro section_func_mwdt, section, symbol
-       .section .\&section\&.\ &symbol, "ax"
+       .section .\&section\&.\&section, "ax"
        FUNC_CODE()
        PERFOPT_ALIGN
        symbol :
@@ -30,7 +30,7 @@
 .endm

 .macro section_subsec_func_mwdt, section, subsection, symbol
-       .section .\&section\&.\ &subsection, "ax"
+       .section .\&section\&.\&subsection, "ax"
        PERFOPT_ALIGN
        symbol :
 .endm
```

@evgeniy-paltsev will fix-up it in his tree.

But now [I really hope the last] problem is:
```
[ 97%] Generating isr_tables.c
cd /SCRATCH/abrodkin/Projects/sources/git/zephyrproject/zephyr/build/zephyr && /remote/ru20home1/abrodkin/.local/lib/python3.7/site-packages/cmake/data/bin/cmake -DELF2HEX=/global/apps/mwdt_2020.06/MetaWare/arc/bin/elf2hex -DELF2BIN=/global/apps/mwdt_2020.06/MetaWare/arc/bin/elf2bin -DSTRIP=/global/apps/mwdt_2020.06/MetaWare/arc/bin/stripac -DOBJCOPY=/global/freeware/Linux/2.X/arc_gnu_2020.03-elf32/bin/arc-elf32-objcopy -DINTARGET=elf32-littlearc -DOUTTARGET=binary -DONLY_SECTION=.intList -DINFILE/SCRATCH/abrodkin/Projects/sources/git/zephyrproject/zephyr/build/zephyr/zephyr_prebuilt.elf -DOUTFILEisrList.bin -P /SCRATCH/abrodkin/Projects/sources/git/zephyrproject/zephyr/cmake/bintools/arcmwdt/elfconvert_command.cmake
Parse error in command line argument: -DINFILE/SCRATCH/abrodkin/Projects/sources/git/zephyrproject/zephyr/build/zephyr/zephyr_prebuilt.elf
Should be: VAR:type=value
```

Note this part `-DOUTFILEisrList.bin` - I guess it should be `-DOUTFILE=isrList.bin`. Any ideas why `=` is not there?

abrodkin commented:
And right as I expected - it was the last one.
With:
```diff
diff --git a/cmake/bintools/arcmwdt/target_bintools.cmake b/cmake/bintools/arcmwdt/target_bintools.cmake
index 7b584585c2..7c3fbadb06 100644
--- a/cmake/bintools/arcmwdt/target_bintools.cmake
+++ b/cmake/bintools/arcmwdt/target_bintools.cmake
@@ -39,8 +39,8 @@ set_property(TARGET bintools PROPERTY elfconvert_flag_section_rename "-DRENAME_S
 set_property(TARGET bintools PROPERTY elfconvert_flag_gapfill "-DGAP_FILL=")
 set_property(TARGET bintools PROPERTY elfconvert_flag_srec_len "-DSREC_LEN=")

-set_property(TARGET bintools PROPERTY elfconvert_flag_infile "-DINFILE")
-set_property(TARGET bintools PROPERTY elfconvert_flag_outfile "-DOUTFILE")
+set_property(TARGET bintools PROPERTY elfconvert_flag_infile "-DINFILE=")
+set_property(TARGET bintools PROPERTY elfconvert_flag_outfile "-DOUTFILE=")

 # - disassembly : Name of command for disassembly of files
 #                 In this implementation `elfdumpac` is used
```

The first test is built!

evgeniy-paltsev commented:
@carlescufi @MaureenHelm
I've pushed PR with all required fixes to work on top of toolchain_abstraction which is ready for merge.

NOTE: now we have several checkpatch errors, however we have them due to checkpatch incompatibility with MWDT assembler macros syntax, so that souldn't be an issue:

```
-:2669: ERROR:SPACING: space required before that '&' (ctx:VxV)
#2669: FILE: include/toolchain/mwdt.h:20:
+	.section .\&section\&.\&symbol, "aw"
```

nashif commented:
> now we have several checkpatch errors,

those can be ignored

tyrantlizard77 reviewed (commented):
OK with merging this but I do ask that enhancement GH tickets are filed for the duplicated linker scripts and linked here before I will +1 this.

Side note: as I see it, the advantage of custom toolchains is improved code generation. I don't see us winning anything by using the third-party linker or assembler however (maybe I'm missing something). Have we considered a hybrid approach where the separate toolchain is used to generate object files from C, but we use GNU assembler and binutils from the SDK for assembly and linking, saving us from having to do all these ugly things in the asm code and linker scripts?

tejlmand commented:
@evgeniy-paltsev @abrodkin  great work.
@MaureenHelm @carlescufi @nashif  Let's get this in.

tejlmand commented:
> OK with merging this but I do ask that enhancement GH tickets are filed for the duplicated linker scripts and linked here before I will +1 this.

@andrewboie  Done, see https://github.com/zephyrproject-rtos/zephyr/issues/28083

ruuddw reviewed (approved):
Thanks all for the quick last-minute work to change to the new properties approach!

ruuddw commented:
@andrewboie: There are various reasons for proprietary toolchains. You already mentioned quality of generation. Some of our customers simply ask for it because it is what they are used to. We thought about a mixed approach, but there is another important  reason for using MetaWare. For ISO26262 functional safety, the toolchain needs top be certified as well and a special version of MetaWare is. That means users don't have to qualify the toolchain themselves for certification.

andrewboie commented:
> For ISO26262 functional safety, the toolchain needs top be certified as well and a special version of MetaWare is. That means users don't have to qualify the toolchain themselves for certification.

If there was ever a specific version of GAS and LD that were also ISO26262 certified, can you use a combination of tools if they are all individually certified? cc @ceolin @AmberHibberd

tejlmand commented:
@andrewboie another reason for preferring 3rd party toolchains is for C++.
Although the GNU `libstdc++` does have an exception to the normal GPLv3, then not all companies are comfortable with this, and prefers a custom toolchain that provides it's own C++ lib, so they are completely free of GPLv3.
https://gcc.gnu.org/onlinedocs/libstdc++/manual/license.html
> The Code: GPL
>
> The source code is distributed under the GNU General Public License version 3, with the addition under section 7 of an exception described in the “GCC Runtime Library Exception, version 3.1” as follows (or see the file COPYING.RUNTIME):

ruuddw commented:
> If there was ever a specific version of GAS and LD that were also ISO26262 certified, can you use a combination of tools if they are all individually certified?

I believe that would help, but there would still be the obligation to 'proof' the combination works.
